### PR TITLE
Hw3

### PR DESCRIPTION
**Pull request by AlexandrBozhok**

### CODE REVIEW

File: Homework_3/hw3.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+# Написать свой декоратор который будет проверять остаток от деления числа 100 на результат работы функции ниже.
+# Если остаток от деления = 0, вывести сообщение "We are OK!», иначе «Bad news guys, we got {}» остаток от деления.
+import functools
+
+
+def check_division_remainder(func):
+    @functools.wraps(func)
+    def wrapper(number):
+        remainder = 100 % func(number)
+        if remainder == 0:
```
Comment (katevoskresenska) [inline Homework_3/hw3.py:11]: if not remainder:

File: Homework_3/hw3.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+# тогда зарейзить ошибку ValueError (raise ValueError(“string type is not supported”))
+
+
+def check_args_type(func):
+    @functools.wraps(func)
+    def wrapper(arg):
+        if type(arg) == int:
+            print(func(arg))
+        elif type(arg) == str:
+            raise ValueError("string type is not supported")
```
Comment (katevoskresenska) [inline Homework_3/hw3.py:29]: It is better to use isinstance(arg, int), isinstance(arg, str)

File: Homework_3/hw3.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    cache_calls_counter = 0
+    func_call_counter = 0
+    @functools.wraps(func)
+    def wrapper(number):
+        if number in cache.keys():
+            nonlocal cache_calls_counter
+            cache_calls_counter += 1
+            print(f"Used cache with counter = {cache_calls_counter}")
+        else:
+            nonlocal func_call_counter
```
Comment (katevoskresenska) [inline Homework_3/hw3.py:51]: wrapper.cache_calls_counter and wrapper.func_call_counter are more preferable than nonlocal variables as they can be re-used

**Comment by katevoskresenska** [comment]
It would be nice to add some documentation strings to all the functions.

File: Homework_3/hw3.py (hunk 2 above)
Comment (AlexandrBozhok) [reply Homework_3/hw3.py:29]: Не знал, спасибо

File: Homework_3/hw3.py (hunk 1 above)
Comment (AlexandrBozhok) [reply Homework_3/hw3.py:11]: Есть ли какая-то разница?

File: Homework_1/additional_for_homework_1.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
+# 1) Write a function def solution(A) that, given an array A of N integers, returns the smallest positive integer (greater than 0) that does not occur in A.
```
Comment (mmykhailenko) [inline Homework_1/additional_for_homework_1.py:1]: this file should be excluded from this PR

File: Homework_2/hw2.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
+# 1) Write a function that emulates the game "rock, scissors, paper"
```
Comment (mmykhailenko) [inline Homework_2/hw2.py:1]: this file should be excluded from this PR

File: Homework_3/hw3.py (hunk 1 above)
Comment (mmykhailenko) [reply Homework_3/hw3.py:11]: more pythonic way

**Review by mmykhailenko — APPROVED** [review]

**Pull request merged**